Sni support

koshelev opened the pull request:
Depends on https://github.com/twitter/finagle/pull/603. Would appreciate a feedback, if this a right direction and what are possible improvements.

Problem

The server name indication is not supported by theNetty4ServerSslHandler at the moment.

Solution

- Netty4SniHandler is implemented. It delays channelActive event  till server_name is processed and than replaces itself with  SslHandler.
- Netty4ServerSslHandler adds the Netty4SniHandler to the pipeline  if SniParameter is available.

Commit:
finagle-core: Delay service instantiation till TLS handshake finished

Problem

When using finagle-netty4 from the develop branch, peer certificate is not available in the context after successful TLS handshake.
https://github.com/twitter/finagle/issues/600

Solution

- changing ChannelTransport.peerCertificate from val to def
- SslServerConnectHandler delays channelRegistered event until
  handshake success. That means ServerBridge's initChannel will be
  delayed.

Result

Service instantiation delayed till TLS handshake finished

ryanoneill commented:
Hi @koshelev, thank you for the pull request. I wanted to let you know that I have looked at the code, and am going to ask for you to make some changes. I am currently thinking about how to fit this within the structure of not only where the code is now, but where it's going as well. I should have a better answer for you with concrete things to change tomorrow (Friday). Thanks.

koshelev commented:
Hi @ryanoneill. Sure, I see the current code in the PR as just a starting point for a discussion and will be happy to improve it. Looking forward to your suggestions!

koshelev commented:
Hi @ryanoneill, any idea when you will have a time to review this PR?

mosesn commented:
LGTM from my perspective, but @ryanoneill is the ultimate authority here

ryanoneill commented:
Hi @koshelev, sorry for the long delay. I'm going to take another look now and see how we can best incorporate this.

ryanoneill commented:
There's a lot of great work here, thank you. I sincerely apologize for the delay. I hadn't looked much into SNI Matching on the server side within Netty before, so I had some items to look into. 

Some of the changes included in this PR have already made it into Finagle which is great, so thank you. I have some suggestions on how to make further changes which would allow us to accept this PR as part of Finagle. If you're interested in working through them, that's great. If not, I'll probably continue this on as time permits with you getting the credit for it.

In this version, `Netty4SniHandler` is extending [SniHandler](https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/ssl/SniHandler.java), which requires a mapping of type `AsyncMapping[String, SslContext]`. This value is passed in via the Finagle stack param that you added which also turns this functionality on. 

We prefer to limit the exposure of Netty specific types in Finagle APIs, especially after the amount of effort required to update from Netty 3 to Netty 4 in Finagle. Also, this uses Netty's `SslContext`, which we don't expose and only use contained within `Netty4ClientEngineFactory` and `Netty4ServerEngineFactory`. 

If we use Netty's `SniHandler`, we're locked into using that `AsyncMapping[String, SslContext]`, which I think is a non-starter for us. After looking at the implementation of `SniHandler` though, I think we can make everything work well by instead extending [AbstractSniHandler](https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/ssl/AbstractSniHandler.java). `AbstractSniHandler` does all of the protocol work, and it would also allow us to use our own types there.

I would prefer if the mapping were along the lines of `Map[String, SslServerConfiguration]`, and then `Netty4SniHandler` used the `Netty4ServerEngineFactory` along with the specific `SslServerConfiguration` to create an `Engine` that's used with the `SslHandler`. That would be more in line with how engine creation and configuration works in Finagle today.

We can then expand on that in the future beyond `Map[String, SslServerConfiguration]` to incorporate different `SslServerEngineFactory` values if necessary.

Would that work for you for what you're trying to do? Let me know your thoughts when you have an opportunity. Thanks again.

ryanoneill reviewed (changes requested):
Changes requested via comment on the pull request.

koshelev commented:
@ryanoneill Thanks a lot for your feedback! I'm on vacation now and will be able to work on the PR again next week.

ryanoneill commented:
@koshelev Hope you're enjoying your vacation! Glad to hear that you're interested in pursuing this.

nepthar commented:
@koshelev Have you had a chance to take a look at this recently?

koshelev commented:
@netphar Unfortunately found no time :(. Will try to find time this or next week.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/twitter/finagle?pullRequest=607) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/twitter/finagle?pullRequest=607) before we can accept your contribution.<br/><hr/>**Vladimir Koshelev** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/twitter/finagle?pullRequest=607) it.</sub>

koshelev commented:
@ryanoneill I finally found some time to work on this PR again. Could you please have a look and let me know WDYT?

ryanoneill commented:
@koshelev I'm going to take another look at this now. Thanks for following up.

koshelev commented:
@ryanoneill have you had a chance to check the PR?

ryanoneill commented:
@koshelev I have not, I'm sorry. I'm going to take a thorough look now.

ryanoneill commented on finagle-core/src/main/scala/com/twitter/finagle/ssl/KeyCredentials.scala:
```diff
@@ -48,4 +50,17 @@ object KeyCredentials {
+  case class KeyAndCertChain(
+      key: PrivateKey,
+      keyCertChain: Array[X509Certificate])
+    extends KeyCredentials
```
Undo this change please. If you would like to add this outside the context of this pull request, we can discuss it in a separate PR.

ryanoneill commented on finagle-core/src/main/scala/com/twitter/finagle/ssl/SslConfigurations.scala:
```diff
@@ -28,6 +31,13 @@ private[ssl] object SslConfigurations {
           case Return(kms) => Some(kms)
           case Throw(ex) => throw SslConfigurationException(ex.getMessage, ex)
         }
+      case KeyCredentials.KeyAndCertChain(key, certChain) =>
```
Undo this change please. If you would like to add this outside the context of this pull request, we can discuss it in a separate PR.

ryanoneill commented on finagle-core/src/main/scala/com/twitter/finagle/ssl/SslConfigurations.scala:
```diff
@@ -58,6 +68,12 @@ private[ssl] object SslConfigurations {
           case Return(tms) => Some(tms)
           case Throw(ex) => throw SslConfigurationException(ex.getMessage, ex)
         }
+      case TrustCredentials.Certificates(certs) =>
```
Undo this change please. If you would like to add this outside the context of this pull request, we can discuss it in a separate PR.

ryanoneill commented on finagle-core/src/main/scala/com/twitter/finagle/ssl/TrustCredentials.scala:
```diff
@@ -34,4 +35,12 @@ object TrustCredentials {
+    *
+    * @certs a collection of X.509 certificates
+    */
+  case class Certificates(certs: Array[X509Certificate]) extends TrustCredentials
```
Undo this change please. If you would like to add this outside the context of this pull request, we can discuss it in a separate PR.

ryanoneill commented on finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/server/Netty4ServerSslHandler.scala:
```diff
@@ -8,9 +13,74 @@ import com.twitter.finagle.ssl.server.{
+import io.netty.channel._
+import io.netty.handler.ssl._
+
+final case class SniSupport(mapping: SniSupport.ServerNameToContext)
```
Please move this to a file named `SniSupport.scala` in the same directory.

ryanoneill replied:
For `mapping`, I believe we should use `Option[Map[String, SslServerConfiguration]]` instead.
None means the parameter is off.
Some(sniMapping) is a mapping from hostnames to SslServerConfigurations

koshelev replied:
`String => TwitterFuture[Option[SslServerConfiguration]]` is more flexible, IMO. `Option[Map[String, SslServerConfiguration]]` implies the mapping is static and keys and certificates are available locally. For instance, it is not true in our case. We have keys and certificates stored in [Vault](https://www.vaultproject.io/) and new servers can be added at any time. That is why I also added https://github.com/twitter/finagle/pull/607/files/61cd8d956037e0efadcb186e48a54790b2c55802#diff-7be4dc145c09c1df2647a7ee1cc8bccd

koshelev replied:
@ryanoneill WDYT?

ryanoneill commented on finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/server/Netty4ServerSslHandler.scala:
```diff
@@ -8,9 +13,74 @@ import com.twitter.finagle.ssl.server.{
+
+object SniSupport {
+
+  type ServerNameToContext = String => TwitterFuture[Option[SslServerConfiguration]]
```
Remove

ryanoneill commented on finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/server/Netty4ServerSslHandler.scala:
```diff
@@ -8,9 +13,74 @@ import com.twitter.finagle.ssl.server.{
+
+  type ServerNameToContext = String => TwitterFuture[Option[SslServerConfiguration]]
+
+  implicit val param: Stack.Param[SniSupport] = Stack.Param(SniSupport( (_: String) => {
```
implicit val param: Stack.Param[SniSupport] = Stack.Param(SniSupport(None))

ryanoneill commented on finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/server/Netty4ServerSslHandler.scala:
```diff
@@ -8,9 +13,74 @@ import com.twitter.finagle.ssl.server.{
+    * Uses JDK SslContext to create an engine rejecting all host names. Is needed to
+    * send the 'unrecognized_name' alert to the client.
+    */
+  val DenialSslContext = {
```
private[ssl]

ryanoneill commented on finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/server/Netty4ServerSslHandler.scala:
```diff
@@ -8,9 +13,74 @@ import com.twitter.finagle.ssl.server.{
+
+  }
+
+  def fromOption(mapping: String => Option[SslServerConfiguration]): SniSupport = SniSupport((serverName: String) => {
```
Remove

ryanoneill commented on finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/server/Netty4ServerSslHandler.scala:
```diff
@@ -74,6 +144,12 @@ private[netty4] class Netty4ServerSslHandler(params: Stack.Params)
     pipeline.addFirst("ssl", sslHandler)
   }
 
+  private[this] def addSniHandler(channel: Channel, factory: SslServerEngineFactory): Unit = {
```
private[this] def addSniHandler(
      channel: Channel,
      factory: SslServerEngineFactory,
      sniMapping: Map[String, SslServerConfiguration]
    ): Unit = {
      ...
    }

ryanoneill commented on finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/server/Netty4ServerSslHandler.scala:
```diff
@@ -84,11 +160,15 @@ private[netty4] class Netty4ServerSslHandler(params: Stack.Params)
-      val sslHandler: SslHandler = createSslHandler(engine)
-      val sslConnectHandler: SslServerConnectHandler = createSslConnectHandler(sslHandler, combined)
-      addHandlersToPipeline(ch.pipeline(), sslHandler, sslConnectHandler)
+      if( params.contains[SniSupport] ) {
```
val SniSupport(mapping) = params[SniSupport]

    mapping match {
      case Some(sniMapping) =>
        addSniHandler(ch, factory, sniMapping)
      case None => 
        ...
    }

ryanoneill commented on finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/server/Netty4SniHandler.scala:
```diff
@@ -0,0 +1,75 @@
+import io.netty.handler.codec.DecoderException
+import io.netty.handler.ssl.{AbstractSniHandler, SniHandler, SslContext, SslHandler}
+import io.netty.util.{AsyncMapping, ReferenceCountUtil}
+import io.netty.util.concurrent.{Future, Promise}
```
Please make this Future => NettyFuture. Since it's the Twitter codebase, Future within it almost always means Twitter Future.

ryanoneill commented on finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/server/Netty4SniHandler.scala:
```diff
@@ -0,0 +1,75 @@
+  * Delays channelActive event till ClientHello with SNI extension is processed, replaces itself with a SslHandler
+  * to continue the handshake.
+  */
+class Netty4SniHandler(mapping: SniSupport.ServerNameToContext, factory: SslServerEngineFactory,
```
mapping: Map[String, SslServerConfiguration]

ryanoneill commented on finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/server/Netty4SniHandler.scala:
```diff
@@ -0,0 +1,75 @@
+
+  override def lookup(ctx: ChannelHandlerContext, hostname: String): Future[Option[SslServerConfiguration]] = {
+    val promise = ctx.executor().newPromise[Option[SslServerConfiguration]]
+    mapping(hostname).respond {
```
promise.setSuccess(mapping.get(hostname))

ryanoneill commented on finagle-netty4/src/test/scala/com/twitter/finagle/netty4/Netty4SslTest.scala:
```diff
@@ -106,33 +98,110 @@ class Netty4SslTest extends FunSuite with Eventually with IntegrationPatience {
+
+    private def mkConfig(serverCert: SelfSignedCertificate): SslServerConfiguration = {
+      SslServerConfiguration(
+        keyCredentials = KeyCredentials.KeyAndCertChain(serverCert.key(), Array(serverCert.cert())),
```
keyCredentials = KeyCredentials.CertAndKey(serverCert.certificate(), serverCert.privateKey()),
    trustCredentials = TrustCredentials.CertCollection(clientCert.certificate())

ryanoneill commented on finagle-netty4/src/test/scala/com/twitter/finagle/netty4/Netty4SslTest.scala:
```diff
@@ -106,33 +98,110 @@ class Netty4SslTest extends FunSuite with Eventually with IntegrationPatience {
+
+    override def server: ListeningServer =
+      mkServer(Params.empty + ServerSsl(Some(SslServerConfiguration(clientAuth = ClientAuth.Needed))) +
+        SniSupport.fromOption(serverCerts.get(_).map(mkConfig)))
```
SniSupport(Some(serverCerts.mapValues(mkConfig)))

ryanoneill commented on finagle-netty4/src/test/scala/com/twitter/finagle/netty4/Netty4SslTest.scala:
```diff
@@ -106,33 +98,110 @@ class Netty4SslTest extends FunSuite with Eventually with IntegrationPatience {
+
+    override def client: Service[String, String] = {
+      val (hostname, serverCrt) = serverCerts.head
+      mkClient(serverCrt.cert(), Some(hostname))
```
Change mkClient to just take hostname: String instead of hostname: Option[String] here, so that it's not necessary to calling hostname.get later.

ryanoneill commented on finagle-netty4/src/test/scala/com/twitter/finagle/netty4/Netty4SslTest.scala:
```diff
@@ -106,33 +98,110 @@ class Netty4SslTest extends FunSuite with Eventually with IntegrationPatience {
       withFixture(test.toNoArgTest(ctx))
     } finally {
-      Await.ready(ctx.close(), 3.seconds)
+      Await.ready(ctx.close(), 30.seconds)
```
30 seconds seems much too long, please reduce.

ryanoneill commented on finagle-netty4/src/test/scala/com/twitter/finagle/netty4/Netty4SslTest.scala:
```diff
@@ -106,33 +98,110 @@ class Netty4SslTest extends FunSuite with Eventually with IntegrationPatience {
+
   test("Peer certificate is available to service") { ctx =>
     val response = Await.result(ctx.client("security is overrated!\n"), 3.seconds)
+    assert(response == "OK\n")
```
This test failed for me until I removed the newline.

ryanoneill commented on finagle-netty4/src/test/scala/com/twitter/finagle/netty4/Netty4SslTest.scala:
```diff
@@ -106,33 +98,110 @@ class Netty4SslTest extends FunSuite with Eventually with IntegrationPatience {
+  override def mkContext: Netty4SslTest.SniCtx = new Netty4SslTest.SniCtx
+
+  test("server uses certificate assigned to the server name sent by client") { ctx =>
+    val response = Await.result(ctx.client("security is overrated!\n"), 300.seconds)
```
Please change from 300 seconds to something more reasonable.

ryanoneill reviewed (changes requested):
Thank you for your perseverance here @koshelev. I feel like this is really close. With the changes suggested, I believe we would be ready to pull this in and put it up for review internally.

Commit:
finagle-netty4: support SNI TLS extension on server

Problem

The server name indication is not supported by the
Netty4ServerSslHandler at the moment.

Solution

- Netty4SniHandler is implemented. It delays channelActive event
  till server_name is processed and than replaces itself with
  SslHandler.
- Netty4ServerSslHandler adds the Netty4SniHandler to the pipeline
  if SniParameter is available.

adleong commented:
What is the status of this PR?  I'm very excited about SNI support in Finagle, is this still being worked on?

bryce-anderson commented:
I'm closing this because it's been inactive for nearly a year. From the history, it looks like it was really close and may be worth picking back up. If that is the case, feel free to reopen this and get it merged with `develop`.
